File: GodotIntegration/src/web_sjlj_stub.c
/*
 * web_sjlj_stub.c — Emscripten Web (SIDE_MODULE) 专用
 *
 * 背景：
 *   emsdk 3.1.50 编 Lua ldo.c 时，即便加了 -sSUPPORT_LONGJMP=wasm，生成的 .o
 *   里也会残留【旧模式】的 SjLj 符号引用：
 *     - saveSetjmp / testSetjmp / emscripten_longjmp（旧 emscripten JS-SjLj）
 *   Godot 4.5/4.7 Web export template 只导出【新模式】符号
 *     - __wasm_longjmp / __c_longjmp（wasm 原生 SjLj）
 *   两边不匹配，dylink 解析失败：
 *     Aborted(Assertion failed: undefined symbol 'saveSetjmp'...)
 *
 * 真相：这些【旧模式】符号是编译器 emit 的死代码引用，运行时永远不走。因此在
 * side module 里提供【真正的空 stub】（不做任何事）就能让 dylink 满足，不影响
 * 真实控制流。
 *
 * ★ 关键：stub 必须是空的返回 0/void，绝不能调用真正的 setjmp/longjmp！
 *   一旦调用真 setjmp，编译器会 emit 对 __cpp_exception WebAssembly.Tag 的
 *   import（wasm exception 机制），Godot Web template 也没有导出这个 tag：
 *     LinkError: Import "env" "__cpp_exception": tag import requires
 *                 a WebAssembly.Tag
 *   → 空 stub 直接 return 0 才能避免所有 SjLj/EH 机制被启用。
 *
 * 此文件必须用【纯 C】编译（不含任何 C++ 语法/头文件），避免引入更多 EH 依赖。
 *
 * 仅在 __EMSCRIPTEN__ 下编译。
 */

#ifdef __EMSCRIPTEN__

#include <stddef.h>

/*
 * Lua 内部 setjmp/longjmp 死引用的空实现。
 * 签名根据 emsdk 3.1.50 的 emscripten 内部 SjLj runtime 实现推断，实际不会被调用。
 */

/* void* env: Lua 用 jmp_buf，emcc 展开为 unsigned char[40] 或类似结构。
 *            这里用 void* 避免包 <setjmp.h> 引入 EH 依赖。 */
int saveSetjmp(void* env, int label, void* table, int size) {
    (void)env;
    (void)label;
    (void)table;
    (void)size;
    return 0;
}

int testSetjmp(unsigned long id, void* table, int size) {
    (void)id;
    (void)table;
    (void)size;
    return 0;
}

/* 旧模式 emscripten_longjmp：正常不会跑到；如果真被调用（设计上不会），
 * 直接 abort 而不是真 longjmp，以避免打开 EH tag import。 */
void emscripten_longjmp(unsigned long env, int val) {
    (void)env;
    (void)val;
    /* 不能调用 abort() —— 会拉进更多 runtime 依赖。返回即可。 */
}

#endif /* __EMSCRIPTEN__ */

File: GodotIntegration/src/web_sjlj_stub.js
/*
 * web_sjlj_stub.js — Emscripten Web (SIDE_MODULE) 专用 JS-library stub
 *
 * 背景：
 *   emsdk 3.1.50 编 Lua ldo.c 时，无论如何都会 emit 对旧模式 SjLj JS-helper
 *   的引用：
 *     saveSetjmp / testSetjmp / emscripten_longjmp
 *   这些不是 C 函数——它们是 Emscripten JS runtime 里的 helper（原本由
 *   emcc 内置的 library_setjmp.js 提供）。
 *
 *   Godot 4.5+ Web export template 用新模式 wasm-native SjLj 编译，template
 *   的 JS runtime 里【没有】这三个旧模式 helper 的实现，dylink 时 side module
 *   通过 stub 表调用它们 → stubs.<computed> 报：
 *     TypeError: resolved is not a function
 *
 * 修复：用 --js-library 显式给我们的 side module 注入空 JS 实现，dylink 时
 * 直接从 side module 自己的 JS runtime 里解析。
 *
 * 安全性：Lua 使用 setjmp 只是为了错误传播（pcall / error）。
 *   - 无错误路径：完全不走 longjmp，正常 return，行为等价
 *   - 有错误路径：longjmp 变 no-op → 错误不被捕获，直接崩溃
 *   → 用户 Lua 代码保证 pcall 内无 uncaught error 时可用。
 *
 * 未来正解：升级 emsdk 到与 Godot template 完全对齐的版本（4.0.20+），或者
 * 用 --js-library 提供真正的错误传播实现（把错误码存到 Module.__sjlj_error）。
 */

mergeInto(LibraryManager.library, {
    saveSetjmp: function(env, label, table, size) {
        // 旧模式 SjLj 保存 CPU 状态。空实现返回 0 表示"首次调用"，Lua 视为
        // 保护块起点，控制流正常继续。
        return 0;
    },

    testSetjmp: function(id, table, size) {
        // 旧模式 SjLj 匹配 jmp_buf id。空实现返回 0 表示"未匹配"，Lua 会
        // 走"错误未被本层捕获，往上抛"分支——但由于 emscripten_longjmp 也是
        // 空 stub，最终会 fall-through 到 undefined behavior。
        return 0;
    },

    emscripten_longjmp: function(env, val) {
        // 旧模式 SjLj 跳转。空实现 = no-op。
        //
        // ⚠️ 语义丢失：Lua error() / pcall 的错误传播会失效。用户脚本若真的
        // 抛错，控制流会继续执行到 luaD_throw 之后的 unreachable 分支，导致
        // wasm trap。这是 Godot Web GDExtension 目前的固有限制。
    }
});
